src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -14,59 +12,82 @@
-        saveProduct(new Product("simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
-        saveProduct(new Product("simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
-        saveProduct(new Product("simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
-        lock.writeLock().unlock();
+    public static ArrayListProductDao getInstance() {
+        ArrayListProductDao result = instance;
+        if (result != null) {
+            return result;
+        }
+        synchronized (ArrayListProductDao.class) {
+            if (instance == null) {
+                instance = new ArrayListProductDao();
+            }
+            return instance;
+        }
     }
 
     @Override
-    public Product getProduct(Long id) throws NoSuchElementException {
+    public Product getProduct(Long id) throws ProductNotFindException {
         lock.readLock().lock();
         try {
             return result.stream().
                     filter(product -> product.getId().equals(id)).
                     findAny().
-                    orElseThrow(NoSuchElementException::new);
+                    orElseThrow(() -> new ProductNotFindException("There is no product with " + id + " id"));
         } finally {
             lock.readLock().unlock();
         }
     }
 
     @Override
-    public List<Product> findProducts() {
+    public List<Product> findProducts(List<String> searchTextList, SortField sortField, SortOrder sortOrder) {
         lock.readLock().lock();
         try {
-            return result.stream().
+            List<Product> products = result.stream().
                     filter(product -> product.getPrice() != null).
                     filter(product -> product.getStock() > 0).
                     collect(Collectors.toList());
+            products = getSearchProduct(products, searchTextList);
+            return sortProducts(products, sortField, sortOrder);
         } finally {
             lock.readLock().unlock();
         }
     }
 
+    private List<Product> getSearchProduct(List<Product> products, List<String> searchTextList) {
+        return products.stream().
+                filter(product ->
+                        searchTextList == null ||
+                                searchTextList.isEmpty() ||
+                                searchTextList.stream().anyMatch(searchText ->
+                                        product.getDescription().toLowerCase().contains(searchText.toLowerCase()))).
```

Comment:
Can we extract these conditions in a separate method?

```
searchTextList == null ||
                                searchTextList.isEmpty() ||
                                searchTextList.stream().anyMatch(searchText ->
                                        product.getDescription().toLowerCase().contains(searchText.toLowerCase()))
```

so the code is something like which is more readable:

`filter(product -> shouldDisplayProduct(searchTextList, product))`